### PR TITLE
NMS-10264: Enhancement to use single alarm to track a problem state

### PR DESCRIPTION
**Pull request by pbrane**
* JIRA: http://issues.opennms.org/browse/NMS-10264

### CODE REVIEW

**Comment by pbrane** [comment]
Ran these tests locally and they pass.

File: opennms-base-assembly/src/main/filtered/etc/events/A10.AX.events.xml [hunk 1]
```diff
@@ -108,8 +108,10 @@
       <descr>Indicates that an SLB virtual service port has come up.</descr>
       <logmsg dest="logndisplay">&lt;p> A10 trap received Description=%parm[#1]% VirtualServer=%parm[#2]%
             Port=%parm[#3]% Type=%parm[#4]%&lt;/p></logmsg>
-      <severity>Normal</severity>
-      <alarm-data reduction-key="%uei%:%parm[#2]%:%parm[#3]%:%parm[#4]%" alarm-type="2" clear-key="uei.opennms.org/traps/A10/axVirtualServerPortDown:%parm[#2]%:%parm[#3]%:%parm[#4]%" auto-clean="false"/>
+      <severity>Cleared</severity>
+      <alarm-data alarm-type="2" reduction-key="uei.opennms.org/traps/A10/axVirtualServerPortDown:%parm[#2]%:%parm[#3]%:%parm[#4]%" auto-clean="false">
+        <update-field field-name="severity" update-on-reduction="true"/>
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/events/A10.AX.events.xml]: What do you think about updating the severity on reduction by default? Would save having to set it for all of these...

**Comment by pbrane** [comment]
> On Jul 27, 2018, at 18:07, Jesse White <notifications@github.com> wrote:
> 
> What do you think about updating the severity on reduction by default? Would save having to set it for all of these...

By default on the clear would be okay I think. I had considered it.  Thanks for the affirmation :)

**Comment by j-white** [comment]
It's green! Updates looks good.

**Comment by pbrane** [comment]
Where is the proper place for me to document the new behavior?

File: opennms-doc/guide-admin/src/asciidoc/text/bsm/introduction.adoc [hunk 2]
```diff
@@ -2,6 +2,7 @@
 // Allow GitHub image rendering
 :imagesdir: ../../../images
 
+[[ga-bsm-introcution]]
```
Comment (j-white) [inline opennms-doc/guide-admin/src/asciidoc/text/bsm/introduction.adoc]: `introcution` vs `introduction`

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 3]
```diff
@@ -117,70 +117,100 @@ public OnmsAlarm persist(Event event) {
...
             throw new IllegalStateException("Event with id " + event.getDbid() + " was deleted before we could retrieve it and create an alarm.");
         }
 
         final String reductionKey = event.getAlarmData().getReductionKey();
         LOG.debug("addOrReduceEventAsAlarm: looking for existing reduction key: {}", reductionKey);
         OnmsAlarm alarm = m_alarmDao.findByReductionKey(reductionKey);
+        
 
-        if (alarm == null) {
+        boolean newIfCleared = Boolean.getBoolean("opennms.alarmd.newIfClearedAlarmExists");
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: This should be declared as a static constant instead of re-reading the system property on every function call - it won't change during the lifetime of the JVM.

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 4]
```diff
@@ -117,70 +117,100 @@ public OnmsAlarm persist(Event event) {
...
+        LOG.debug("addOrReduceEventAsAlarm: reduceClearedAlarms is {}.", newIfCleared);
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: This can be printed in the constructor instead of on every call to `addOrReduceEventAsAlarm`.

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 5]
```diff
@@ -117,70 +117,100 @@ public OnmsAlarm persist(Event event) {
...
+        
+        if (alarm == null || (alarm.getSeverity().compareTo(OnmsSeverity.CLEARED)==0 && newIfCleared)) {
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: I'd swap `newIfCleared` to the left of the `&&` since it's faster to check and likely to be `false` (default).

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 6]
```diff
@@ -117,70 +117,100 @@ public OnmsAlarm persist(Event event) {
...
             if (LOG.isDebugEnabled()) {
-                LOG.debug("addOrReduceEventAsAlarm: reductionKey:{} not found, instantiating new alarm", reductionKey);
+                LOG.debug("addOrReduceEventAsAlarm: reductionKey:{}, instantiating new alarm", reductionKey);
             }
-            alarm = createNewAlarm(e, event);
+            
+            if (alarm != null) {
+                LOG.debug("addOrReduceEventAsAlarm: altering reductionKey of currently cleared Alarm for problem: {}; A new Alarm will be instantiated to manage problem.", reductionKey);
+                String uniqueReductionKey = alarm.getReductionKey() + ":ID:" + alarm.getId().toString();
+                LOG.debug("addOrReduceEventAsAlarm: "+uniqueReductionKey);
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: We should use `{}` instead of concatenating the strings here.

File: opennms-base-assembly/src/main/filtered/etc/opennms.properties [hunk 7]
```diff
@@ -171,6 +171,13 @@ rrd.binary=${install.rrdtool.bin}
 # Default: false
 #org.opennms.utils.propertiesCache.enableCheckFileModified=false
 
+###### Alarmd Properties ######
+#
+# Disable this property to force Alarmd to create new alarms when an problem re-occurs and the
+# existing Alarm is in a "Cleared" state.
+#
+# Default: false
+#opennms.alarmd.newIfClearedAlarmExists="false"
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/opennms.properties]: Double quotes aren't needed here.

**Comment by RangerRick** [comment]
I went ahead and merged develop one last time to fix the conflict.  Assuming this passes we can merge it.

**Comment by j-white** [comment]
Are we ready to merge this, or are we still adding features here? We're now missing test coverage for the new flag.

**Comment by pbrane** [comment]
Following today's comments, just trying to make this so that it's less impactful on the community.

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java [hunk 8]
```diff
@@ -548,9 +545,86 @@ public void testArchiveAlarm() throws Exception {
...
+        await().atMost(10, SECONDS).until(numAlarmsCallable, equalTo(1));
+
+        // Send in the UP
+        sendNodeUpEvent(node);
+
+        // We should have two alarms now
+        numAlarmsCallable = getNumAlarmsCallable();
+        await().atMost(10, SECONDS).until(numAlarmsCallable, equalTo(2));
+
+        Thread.sleep(1000);
```
Comment (j-white) [inline opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: If we're waiting for the alarm to be cleared, we should use `await()....` with that condition instead of an explicit sleep.

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java [hunk 9]
```diff
@@ -548,9 +545,86 @@ public void testArchiveAlarm() throws Exception {
...
+
+        final MockNode node = m_mockNetwork.getNode(1);
+
+        // There should be no alarms in the alarms table
+        assertEmptyAlarmTable();
+
+        //this should be the first occurrence of this alarm
+        //there should be 1 alarm now
+        String dbname = m_database.getTestDatabase();
+        System.out.println(dbname);
```
Comment (j-white) [inline opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: ?

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java [hunk 10]
```diff
@@ -548,9 +545,86 @@ public void testArchiveAlarm() throws Exception {
...
```
Comment (j-white) [inline opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: ^

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java [hunk 11]
```diff
@@ -548,9 +545,86 @@ public void testArchiveAlarm() throws Exception {
...
+        await().atMost(10, SECONDS).until(numAlarmsCallable, equalTo(1));
+
+        // One alarm should be cleared, and not archived
+        assertThat(m_alarmDao.findAll().stream().filter(a -> !a.isArchived()
+                && OnmsSeverity.CLEARED.equals(a.getSeverity())).count(), equalTo(1L));
+
+    }
 
     private Callable<Integer> getNumAlarmsCallable() {
-        return () -> m_alarmDao.countAll();
```
Comment (j-white) [inline opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: Unnecessary change.

File: opennms-base-assembly/src/main/filtered/etc/events/opennms.alarm.events.xml [hunk 12]
```diff
@@ -19,12 +19,4 @@
          <update-field field-name="severity" update-on-reduction="true"/>
       </alarm-data>
    </event>
-   <event>
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/events/opennms.alarm.events.xml]: Can we restore this?

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java [hunk 13]
```diff
@@ -122,15 +125,24 @@ public OnmsAlarm persist(Event event) {
...
         }
 
         final String reductionKey = event.getAlarmData().getReductionKey();
         LOG.debug("addOrReduceEventAsAlarm: looking for existing reduction key: {}", reductionKey);
-        OnmsAlarm alarm = m_alarmDao.findByReductionKey(reductionKey);
+        
+        String key = reductionKey;
+        String clearKey = event.getAlarmData().getClearKey();
+        
+        if (!m_legacyAlarmState && clearKey != null && event.getAlarmData().getAlarmType() == 2) {
```
Comment (j-white) [inline opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: We should use the `OnmsAlarm.RESOLUTION_TYPE` constant, or the `isResolutionEvent` utility method you've added here.

File: integration-tests/config/src/test/java/org/opennms/netmgt/config/WillItUnmarshalIT.java [hunk 14]
```diff
@@ -310,6 +310,67 @@ private static void addFile(final Source source, final String file, final Class<
         addFile(Source.EXAMPLE, "thresholds.xml", ThresholdingConfig.class, false, null);
         addFile(Source.EXAMPLE, "tl1d-configuration.xml", Tl1dConfiguration.class, false, null);
         addFile(Source.EXAMPLE, "viewsdisplay.xml", Viewinfo.class, false, null);
+        addFile(Source.EXAMPLE, "events-with-legacy-alarm-data/A10.AX.events.xml", Events.class, true, null);
```
[the remark this answers is not in the document]
Comment (pbrane) [reply integration-tests/config/src/test/java/org/opennms/netmgt/config/WillItUnmarshalIT.java]: √ (I actually thought of these, yesterday, and forgot 'bout them, thanks!

File: opennms-doc/guide-admin/src/asciidoc/text/alarms/introduction.adoc [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+.Both Service Down Events and the previous Service restored Event from the Poller
+image:alarms/single_alarm_6.png[]
+
+.The Alarm is again cleared immediately (notice counter doesn't increment)
+image:alarms/single_alarm_7.png[]
+
+.Both Service Down and restored Events
+image:alarms/single_alarm_8.png[]
+
+The previous design of having two Alarms to track the state of problems can be re-enabled by copying the files from the $OPENNMS_ETC/examples/events-with-legacy-alarm-data/ folder to the $OPENNMS_ETC/events/ folder.
```
[the remark this answers is not in the document]
Comment (pbrane) [reply opennms-doc/guide-admin/src/asciidoc/text/alarms/introduction.adoc]: √ Thanks for being so detailed oriented on this.  I was pretty tired when I made those last doc changes and should have looked at them again, today.

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java (hunk 8 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: I tried... will try that again.

Comment (pbrane) [reply opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: √ thanks for the help.

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java (hunk 9 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: √

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java (hunk 10 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: √

File: opennms-base-assembly/src/main/filtered/etc/events/opennms.alarm.events.xml (hunk 12 above)
Comment (pbrane) [reply opennms-base-assembly/src/main/filtered/etc/events/opennms.alarm.events.xml]: √

File: opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java (hunk 13 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/main/java/org/opennms/netmgt/alarmd/AlarmPersisterImpl.java]: √

File: opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java (hunk 11 above)
Comment (pbrane) [reply opennms-alarms/daemon/src/test/java/org/opennms/netmgt/alarmd/AlarmdIT.java]: √
